core/src/main/java/com/datastax/oss/driver/api/core/ClusterBuilder.java

```diff
@@ -135,14 +160,15 @@ public ClusterBuilder addTypeCodecs(TypeCodec<?>... typeCodecs) {
     Set<InetSocketAddress> contactPoints =
         ContactPoints.merge(programmaticContactPoints, configContactPoints);
 
-    InternalDriverContext context = new DefaultDriverContext(configLoader, typeCodecs);
-    return DefaultCluster.init(context, contactPoints);
+    return DefaultCluster.init((InternalDriverContext) buildContext(configLoader), contactPoints);
   }
 
-  /** Convenience method to call {@link #buildAsync()} and block on the result. */
-  public Cluster<CqlSession> build() {
-    BlockingOperation.checkNotDriverThread();
-    return CompletableFutures.getUninterruptibly(buildAsync());
+  /**
+   * This <b>must</b> return an instance of {@code InternalDriverContext} (it's not expressed
+   * directly in the signature to avoid leaking that type through the protected API).
+   */
+  protected DriverContext buildContext(DriverConfigLoader configLoader) {
```

Comment:
This is a bit disingenuous: if a class using the protected API requires a type, then that type should be in the public API. But we're providing a convenience for custom driver extensions here, which use the internal APIs anyway, so I think the compromise is acceptable.
I've tried extracting a `ClusterBuilder` interface to avoid this problem, but it doesn't work well with self-types (you can't extend the interface and get covariant returns automatically, you'd have to override all methods).